CSM flavor limits: Add option to disable mod loading

paramat opened the pull request:
WIP not tested yet.

Alternative to #6710 . Adds the option for a server to disable loading of client-provided CSMods.
If the new flavour limit 0x80000000 is set, CSMods are not loaded.
The clientside setting remains.
Logic: If either the client or the server disables CSModding, CSMods are not loaded.

Not sure it will work, depends on whether the flavour setting is sent to the client before CSMod loading, i'll soon find out in testing.

If you are a server owner please state your support or otherwise.
@Ezhh @tenplus1 @shivajiva101 @VanessaE @Megaf @ExeterDad

BluebirdGreycoat commented:
I support. With the understanding that it has not been tested yet.

BluebirdGreycoat commented:
I assume it's alright if I ping additional server owners.

@sorcerykid @maikerumine

(If I've accidentally pinged the wrong person, sorry.)

Megaf commented:
@paramat Thanks for this.
I will be testing this myself as soon as I can.

nerzhul commented:
you just break how CSM work if you do this, client will be totally broken. 👎

Megaf commented:
@nerzhul Well, We are open for alternatives. There **must** be a way to completely disable CSM. Server owners/admins must be able to control their own server.

So, you have any other idea?

tenplus1 commented:
We really need a CSM kill switch for the client when connecting to specific servers.

Megaf commented:
I see this is not going to be merged then, continuing discussion on https://github.com/minetest/minetest/issues/6699

t0ny2 commented:
Instead of comparing a bitfield to an integer, why not use bit 63 to indicate CSM disabled/enabled. That way you get a kill switch using the existing flavours system without breaking how flavours currently work.

nerzhul commented:
@t0ny2 it's reasonable yes, use the last byte for that is a nice idea, call it FLAVOUR_DISABLE_MOD_LOADING

ExeterDad commented:
Since you pinged me. 
I have already edited what was going to be a long winded, and mostly a rant of a post.
I'm excited about CSM. But until there is complete "choice" both ways, I support a (optional) kill switch for both client and server. In it's current state, CSM has me wondering if it's actually worth paying hosting for yet another month of frustration and toil.

I'm hoping to see support for some kind of "signed by author" CSM code. To be verified both by client and server, and whitelisted on each side. So the client and server can agree or disagree whether they choose to do business with each other.

shivajiva101 commented:
Both the client and server should be able to disable CSM as they see fit, even if that means losing game features/players. It's more about the right to choose than any other consideration. That way if it turns out there is more holes in it than a jumbo packet of polo mints, we retain effective control.

SmallJoker commented on src/client.cpp:
```diff
@@ -118,8 +118,9 @@ void Client::loadMods()
-	if (!m_modding_enabled) {
+	// If modding is not enabled, don't load mods, just builtin.
+	// 0xFFFFFFFF is CSM flavour setting for 'disable all'.
+	if (!m_modding_enabled || m_csm_flavour_limits == 0xFFFFFFFF) {
```
`|| m_csm_flavour_limits == CSMFlavourLimit::CSM_FL_ALL`
Makes the comment line above superfluous.

paramat replied:
Ok, duh =)

The function won't work it seems, because it would be true if any flavor is used (?).

SmallJoker replied:
Good point, that's true. Go ahead with checking the variable directly then.

paramat replied:
Well i'll do what has been suggested.

paramat commented:
> you just break how CSM work if you do this, client will be totally broken

Hm, Why? It just does the same as using the client setting to disable.

> also you will break a future server sent feature because the whole lua stack is disabled if you want to disable only flavours

In the case of server-provided CSM the server would not set the 'disable all' option, that could be done automatically when server-provided is used. I can't see how anything is broken.

> use the last byte for that is a nice idea, call it FLAVOUR_DISABLE_MOD_LOADING

Yes i can do that, it occured to me. Will do.

sorcerykid commented:
Thanks for the tag, GoldFireUn. As a server operator myself, I'm disappointed by the current implementation of CSM.

There is certainly no reason that end-users should be circumventing the game mechanics of a server. That is not only a blatant disregard for the artistic vision of the subgame designer, but it's an open invitation to abuse and other misconduct.

An optional "disable flag" is easily defeated by anybody that is compiling their own executable. I would suggest that CSM be removed from the Minetest engine until the inherent security flaws can be properly addressed. An alternate stopgap would be to disable CSM altogether for multiplayer, foregoing the proposed flag.

I was really hoping for an implementation of CSM closer to the official design document:

http://dev.minetest.net/Client_scripting_plans

basicer commented:
>  An optional "disable flag" is easily defeated by anybody that is compiling their own executable. 

Could such a person not just as easily `git revert` whatever change you're suggesting?

BluebirdGreycoat commented:
If they're a coder all bets are off. That has always been the case.

The original design document should have been followed, if CSM was to be possible at all.* It's what server operators expected; I certainly did. I do not believe there are any valid excuses for disregarding it.

sorcerykid commented:
> Could such a person not just as easily git revert whatever change you're suggesting?

Perhaps, but if 0.5.0 is going to be a breaking release anyway, then it seems this is the optimal time to ween people off of CSM until a better implementation is deviced. To maintain what is clearly a premature "edition" of CSM across a major version change, only to completely overhaul it in a successive feature update of the engine (such as 0.5.1) isn't particularly sensible. From a development standpoint, it seems that this is the best opportunity to remove it.

paramat commented:
Keep in mind that server-provided CSM has more potential and is still coming, and won't take much more code to complete, according to red-001. I've asked red-001 to work on adding server-provided CSM, and i hope nerzhul does too if able to (he's very busy IRL currently).
So the server-provided aspect will be added as in the original plan.
Also remember that 0.5.0 will improve the situation, as CSM flavors and (hopefully) this PR will be usable.

paramat commented:
Updated as requested, still need to test. Then i'll do docs last.

paramat commented:
Problem, inserting code to print the value of `m_csm_flavour_limit` at this point in the code shows that it has not yet been sent to the client when deciding mod loading.
Any suggestions? Is it possible to send the flavour earlier?

paramat commented:
Maybe move `SendCSMFlavourLimits` from `Init2` to `Server::handleCommand_Init()`?

nerzhul commented:
Flavour are sent in init2 phasis. See SendCSMFlavourLimits(pkt->getPeerId());

This seems to be the good phasis. Mods should be loaded later if we want to make this switch work, as now they are loaded just after connecting (see game.cpp L2148) (builtin included)

I think to solve this you should change the CSM init in two phasis, builtin kept in this model and mod loading after flavour receiving

maikerumine commented:
As a survival server operator and a game designer, I would like the option to shut off CSM.  I know hacked compiled clients will always be here, but CSM makes the cheating problem too easy.

stujones11 commented:
> CSM makes the cheating problem too easy.

What cheating problem? For the last time, ore-detection is the only possible cheat with current CSM and this can be disabled by 'flavours' or however you like to spell that :P

Perhaps the best compromise in the meantime would be to disable CSM in the backport release, assuming there is to be one.

paramat commented:
Thanks nerzhul will attempt.

> ore-detection is the only possible cheat

Don't trivialise that, it's not just ores next to air, you can detect any node anywhere, x-ray texture packs didn't show nodes without overlay textures.
It's not the only cheat, and many seemingly 'harmless' CSM functions can still change gameplay and give players advantages.

> Tip: Don't base your server's economy on mined ores, CSM problem solved :P)

Suggesting servers radically change their gameplay to neutralise cheating is very insulting. It's blaming the victim for the abuse and not taking action against the abuser.
It affects more than an economy based on ores.

stujones11 commented:
> Suggesting servers radically change their gameplay to neutralise cheating is very insulting.

@paramat I fear you are taking yourself and my comment way too seriously, it was merely a suggestion.

BluebirdGreycoat commented:
CSM is a serious issue for all sides.

paramat commented:
No problem if it was a joke, but a suggestion is still insulting =)

paramat commented:
I'll probably need more help with this.

stujones11 commented:
Completely off-topic but IMO basing your server economy exclusively on mined ores is probably _the_ most unimaginative things you could possible do.

paramat commented:
Basing an economy on valuable ores seems obvious and intuitive, although it may have issues.
Since these are deep underground there is little aesthetic impact, ores are there to be mined.
Encouraging exploration, horizontally or vertically is a good thing, which also helps to use the world height.

stujones11 commented:
> Since these are deep underground

This is precisely why CSM ore-detection mod will give you very little game advantage in the grand scheme of things, you still have to dig for to actually ore...

paramat commented:
Yes, but you know precisely where the cluster is, the valuable ores are in clusters spaced far apart. Digging straight to one is a huge difference to excavating a cube equal to the average spacing. That's a huge advantage in time and tool wear, not 'very little'.
Even 'little' advantage is still an advantage that affects a server and so justifies a restriction.

stujones11 commented:
> Yes, but you know precisely where the cluster is

Did you try any of these client mods yourself? If not I'd suggest you give them a go, then decide if you think it was worth the hassle.

paramat commented:
No. They may not have good implementation. I don't need to try them because i know what unrestricted CSM can do, if you can detect certain nodes around you and get a co-ordinate you can dig straight towards them. The average diamond spacing is 16 nodes, the difference in number of nodes needed to be dug per cluster is 3 * 16 = 48 versus 16 ^ 3 = 4096, that's 85 times less time and tool wear.

Ezhh commented:
It doesn't matter if someone thinks something is or is not a big deal. The end result is the same; CSM weakens (too greatly) my ability to decide how playing on my own servers works. That comes before we even get into whether cheating is possible and to what extent that matters.

It should be wonderfully clear by now that server owners are not happy. Not all server owners are unhappy for the exact seem reasons, or even have the exact same concerns about CSM, but the concern about CSM has been made clear.

And telling server owners what amounts to "it's not that bad" / "just change how you do things" won't help . Several of the devs have admitted CSM should not have been added in it's current form, so we should be beyond getting derailed by debates about how bad ore detect is.

nerzhul commented:
just a suggestion to stop the hemoragy, finish this PR properly, backport flavour to 0.4 and release 0.4.17 with some high priority fixes + flavours.

sfan5 commented:
Backporting flavors is not possible unless we want to break backwards compatibility, which is the exact opposite of what 0.4.17 is supposed to be.

paramat commented:
Best we don't backport flavors, we want people to update to 0.5.0 as soon as possible.

@rubenwardy and others i'll probably need help with this PR, i'll attempt it but a little out of my depth.
@nerzhul can you show me at what point 'load mods' should now be?

Ezhh commented:
Flavors not having been released in full with CSM should be classed as a bug. Please don't try and use this as a method of pressuring server owners to update quicker than they are happy to. This is a compatibility breaking release which can have big consequences for servers, so **IF** it's possible to backport this you should.

paramat commented:
The reasons i feel it's best that servers update to 0.5.0 ASAP are explained in #6542 , although note at that time i didn't want a 0.4.17 release at all, but am now neutral.

Ezhh commented:
Feeling that servers should update quickly doesn't make not fixing a problem in a bug-fix release right. If 0.5 is worth it, people will switch regardless of this.

paramat commented:
I've noticed some people (not you Ezhh) have a misunderstanding of 0.5.0, because we call it a 'breaking release' some seem to think lots of things will be broken, only compatibility is broken, mods and worlds are not, apart from player model vertical offset (quickly fixed in blender as i'm sure the mod authors will do quickly before 0.5.0), and the position of attachment of players to vehicles.

So unfortunately there seems to be an unnecessary amount of hesitation to updating to 0.5.0.
Backporting CSM flavours will reduce what 0.5.0 has going for it, adding to this problem.
I know that my approach will be unpopular with some, however i believe it is in the best interest of MT as a whole, and i'm prepared to be a little unpopular to do the right thing.

However it's clear that, as long as the official Android app is improved before 0.5.0 (i think we should delay 0.5.0 until it is good) then moving players to the official app will bring a huge benefit to servers, so it's actually in server owners best interests.
Assuming CSM flavours cannot be backported, which seems likely, 0.5.0 will also bring flavours and my upcoming 'master disable', so it seems updating to 0.5.0 is an advantage.

You have said you are concerned about losing players, assuming the official app is improved in time, how will you lose players when they can use that?

Server owners continuously state how the more unpleasant unofficial apps are a huge problem, along with hacked clients. We now have a 'once in 6 years' chance to move a huge amount of players onto the official app. Because the official app is not hacked this will also reduce cheating clients.
The forked apps will soon catch up with 0.5.0 so every week of delay before updating will reduce this chance.

paramat commented:
Here's a big issue, i know that Multicraft will not be updating to 0.5.0, and it has > 1 million downloads, which seems to be 10 times total official MT downloads. As servers update to 0.5.0 these players will be forced to move to a new app. Mulitcraft is one of the better behaved apps, if we don't provide a good official app they are likely to move to worse ones. If servers update to 0.5.0 quickly there's more chance these players will move to the official app, before the unofficial ones catch up.

paramat commented:
Updated but in testing.

I looked at https://github.com/minetest/minetest/blob/6bab695479d42ac2b051b8a35639c32b22efbf8d/src/clientiface.h#L68
CSMflavour limits are sent in 'INIT2', so i have moved client mod loading from `Game::connectToServer` to `Game::getServerContent` which appears to be after the INIT2 stage.

paramat commented:
Ezhh i moved offtopic discussion to https://github.com/minetest/minetest/issues/6542#issuecomment-348742003

paramat commented:
No, doesn't work, the CSM flavour limit is still 0 at that point in client.cpp and this error occurs:

![screenshot from 2017-12-03 06-12-21](https://user-images.githubusercontent.com/3686677/33522942-4f709624-d7f1-11e7-8682-bac2e599af43.png)

nerzhul commented:
@sfan5 how flavour limits break compat ?

paramat commented:
I just thought of something, since builtin mods will still be loaded, could someone place their mod in the builtin folder and bypass this PR?

BluebirdGreycoat commented:
Isn't it true that the only item of consequence that clientside `builtin` does is handle the death formspec? If so, that could be moved into C++, or remade into a mod that's part of Minetest Game. Then the rest of clientside `builtin` could be disabled if CSM is disabled, without creating any other problems.

nerzhul commented:
@BluebirdGreycoat originally it was in C++ and it was moved to Lua to permit overriding death formspec with mods

BluebirdGreycoat commented:
In that case I would argue that it should be moved out of `builtin` and made part of Minetest Game. (It could still be overridden, there.) Then clientside `builtin` could be safely disabled if the server asks that all CSM be disabled.

paramat commented:
Yes @rubenwardy and myself feel there should be no builtin CSM stuff.

nerzhul commented:
It's mod unloading not csm unloading.I prefer adding builtin code signature as builtin permits to offload some minetest client parts-------- Message original --------Objet : Re: [minetest/minetest] CSM flavor limits: Add option to disable mod loading (#6718)De : Bluebird À : minetest/minetest Cc : Loïc Blot ,Mention In that case I would argue that it should be moved out of builtin and made part of Minetest Game. (It could still be overridden, there.) Then clientside builtin could be safely disabled if the server asks that all CSM be disabled.

—You are receiving this because you were mentioned.Reply to this email directly, view it on GitHub, or mute the thread.


  
  




{"api_version":"1.0","publisher":{"api_key":"05dde50f1d1a384dd78767c55493e4bb","name":"GitHub"},"entity":{"external_key":"github/minetest/minetest","title":"minetest/minetest","subtitle":"GitHub repository","main_image_url":"https://cloud.githubusercontent.com/assets/143418/17495839/a5054eac-5d88-11e6-95fc-7290892c7bb5.png","avatar_image_url":"https://cloud.githubusercontent.com/assets/143418/15842166/7c72db34-2c0b-11e6-9aed-b52498112777.png","action":{"name":"Open in GitHub","url":"https://github.com/minetest/minetest"}},"updates":{"snippets":[{"icon":"PERSON","message":"@BluebirdGreycoat in #6718: In that case I would argue that it should be moved out of `builtin` and made part of Minetest Game. (It could still be overridden, there.) Then clientside `builtin` could be safely disabled if the server asks that all CSM be disabled."}],"action":{"name":"View Pull Request","url":"https://github.com/minetest/minetest/pull/6718#issuecomment-348888710"}}}

paramat commented:
>  since builtin mods will still be loaded, could someone place their mod in the builtin folder and bypass this PR?

Please can someone confirm this? In IRC this seems relevant:

> 10:47 nerzhul          this thing will try to disable mods loading, but if you add code to builtin CSM on various places you can make it work like a rogue mod (like the hated oredetect)

So we should revert the death formspec thing. Having builtin stuff using CSM is not a good idea.

paramat commented:
New issue for this #6736

nerzhul commented:
@paramat the idea is not to block ourselves to use builtin CSM to move hardcoded things to lua, i tend to prefer lua signing

nerzhul commented:
I take the ownership of this PR in #6738 to add more fixes & additions

paramat commented:
Ok thanks, this PR was going nowhere.
